Refactoring/docker from properties

alex-glv opened the pull request:
The PR allows to pass docker configuration from JVM properties:
*  docker.host
*  docker.cert.path

This will resolve issue when running gradle from an IDE, easier testing on docker-machine.

sadovnikov commented:
@alex-glv, could you provide an example of how you'd use this in InteliJ?

alex-glv commented:
The problem:
I am running a docker-machine and docker client needs to know where to find the docker engine.
Usual approach, add environment variables (DOCKER_HOST, etc..), however there's no option to set the env variables for gradle tasks.
For gradle tasks, you can set VM options:
```-Ddocker.host=tcp://192.168.99.101:2376 -Ddocker.cert.path=/home/alg/.docker/machine/machines/default```

Now minimesos can find docker engine without env variables

sadovnikov commented on minimesos/build.gradle:
```diff
@@ -138,22 +138,36 @@ test {
+        certPath = new File(System.properties.getProperty("docker.cert.path"))
+    }
+
+    if (System.getenv("DOCKER_HOST") != null) {
```
Doesn't `if (System.env.DOCKER_HOST) {` have the same result, does it?

sadovnikov commented on minimesos/build.gradle:
```diff
@@ -138,22 +138,36 @@ test {
+        url = System.properties.getProperty("docker.host")
+    }
+
+    if (!System.properties.getProperty("docker.cert.path", "").isEmpty()) {
```
Would `if (System.properties.getProperty("docker.cert.path")) {` do the same?

sadovnikov commented on minimesos/build.gradle:
```diff
@@ -138,22 +138,36 @@ test {
+    }
+
+    if (System.getenv("DOCKER_HOST") != null) {
+        url = System.getenv("DOCKER_HOST")
```
to ease reading, I would move `url = url.replace("tcp", "https")` here

frankscholten commented:
Add a unit test for `DockerClientFactory` to ensure the system properties are set correctly

alex-glv commented:
Thanks for the comments. Added unit tests and refactoring around docker client factory.
Renamed properties to be consistent with DockerClientConfigBuilder (docker.host -> docker.io.url, etc...)

Ready for review.

frankscholten commented on minimesos/src/main/java/com/containersol/minimesos/docker/DockerClientFactory.java:
```diff
@@ -3,29 +3,77 @@
+        if (dockerCertPathEnv != null && new File(dockerCertPathEnv).exists()) {
+            dockerCertPath = dockerCertPathEnv;
+        }
+        builder.withUri(dockerUri.replace("tcp", "https"));
```
It seems that in case of setting the Docker host the properties take precedence while in case of cert path the environment variable takes precedence.

alex-glv replied:
ENV variables take priority in both cases.

frankscholten replied:
If ENV should take prio you have to swap lines 46-47 with 48-50

frankscholten commented on minimesos/src/test/java/com/containersol/minimesos/docker/DockerClientFactoryTest.java:
```diff
@@ -0,0 +1,45 @@
+        props.setProperty("docker.io.dockerCertPath", "/obviously/does/not/exist");
+        DockerClientFactory.build(props);
+        assertEquals(DockerClientFactory.getDockerCertPath(), null);
+    }
```
Also add a test that shows which configuration option, environment variable or property takes precedence.

alex-glv replied:
Good idea, thanks

alex-glv commented:
Added the ENV vs Properties priority test

sadovnikov commented on minimesos/src/main/java/com/containersol/minimesos/docker/DockerClientFactory.java:
```diff
@@ -3,28 +3,84 @@
+    private static String dockerHostEnv;
+    private static String dockerCertPathEnv;
+    private static String dockerHostPrp;
+    private static String dockerCertPathPrp;
```
These four are using only within a single method and therefore can (should) become local variables:
```
    private static String dockerHostEnv;
    private static String dockerCertPathEnv;
    private static String dockerHostPrp;
    private static String dockerCertPathPrp;
```

sadovnikov commented on minimesos/src/main/java/com/containersol/minimesos/docker/DockerClientFactory.java:
```diff
@@ -3,28 +3,84 @@
 
-            DockerClientConfig config = builder.build();
-            dockerClient = DockerClientBuilder.getInstance(config).build();
+    public static DockerClientConfigBuilder populateConfigBuilder(DockerClientConfigBuilder builder) {
```
Do you think this method is supposed to be used outside of the class? If not, it should be marked as `private`

sadovnikov commented on minimesos/src/test/java/com/containersol/minimesos/docker/DockerClientFactoryTest.java:
```diff
@@ -0,0 +1,65 @@
+import static org.junit.Assert.*;
+
+public class DockerClientFactoryTest {
+    String dummyCertPath;
```
please, mark as `private`

sadovnikov commented on minimesos/src/test/java/com/containersol/minimesos/docker/DockerClientFactoryTest.java:
```diff
@@ -0,0 +1,65 @@
+    String dummyCertPath;
+    @Before
+    public void before() throws IOException {
+        File crt = new File("/tmp/" + String.valueOf(Math.round(Math.random()*100)));
```
`/tmp/` does not exist everywhere. Please, use `File.createTempFile()`

sadovnikov commented on minimesos/src/main/java/com/containersol/minimesos/docker/DockerClientFactory.java:
```diff
@@ -3,28 +3,84 @@
-            if (StringUtils.isBlank(dockerHostEnv)) {
-                builder.withUri("unix:///var/run/docker.sock");
-            }
+    public static DockerClient build(Properties argProperties) {
```
Seems, it's either system properties or environment variables, but not a mix of them. Right?

sadovnikov commented on minimesos/src/main/java/com/containersol/minimesos/docker/DockerClientFactory.java:
```diff
@@ -3,28 +3,84 @@
+        return builder;
+    }
+
+    public static DockerClient getDockerClient() {
```
What is the difference between this method and `build()`? Could you JavaDoc all public methods?

frankscholten commented:
Closing for now. Branch is out of date and it does not solve an issue we are currently experiencing.
